### ISSUE TITLE
Component Test with Vite cold-start issue causes first test to fail

### ISSUE THREAD

**Issue by Craig1f** [cypress-io/cypress#22557]
### Current behavior

```
:~/environment/my-app (my-branch) $ npx cypress run --component
...
  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:        10.2.0                                                                         │
  │ Browser:        Electron 100 (headless)                                                        │
  │ Node Version:   v14.16.1 (/home/ec2-user/.nvm/versions/node/v14.16.1/bin/node)                 │
  │ Specs:          6 found (.................................................)                    │
  │ Searched:       **/*.cy.{js,jsx,ts,tsx}                                                        │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
Running:  ProfileCard.cy.js                                                               (1 of 6)
12:23:30 PM [vite] ✨ new dependencies optimized: vuetify/lib
12:23:30 PM [vite] ✨ optimized dependencies changed. reloading


  (Attempt 1 of 3) An uncaught error was detected outside of a test
  (Attempt 2 of 3) An uncaught error was detected outside of a test
  1) An uncaught error was detected outside of a test

  0 passing (2s)
  1 failing

  1) An uncaught error was detected outside of a test:
     TypeError: The following error originated from your test code, not from Cypress.

  > Failed to fetch dynamically imported module: http://localhost:3000/__cypress/src/src/components/ProfileCard.cy.js
...
```

The other tests pass. 

On the second run, tests pass. 

First test always fails in CI pipeline on Github Actions.

Adding test retries does not help

### Desired behavior

Some way of preventing the first test from failing. 

Some way of warming up Vite so that it doesn't have this cold-start problem (I'm also going to file a bug report there). 

### Test code to reproduce

Proprietary code. Will work on setting up a forked repo if necessary. Will take time. 

### Cypress Version

10.2.0

**Comment by Craig1f** [comment]
Note: 

I fixed the CI pipeline in GitHub Actions by adding this before my actual test run: 

```
      - name: Fix for vite cold-start issues with component testing
        continue-on-error: true
        run: npx cypress run --component --spec src/components/random-test-here.cy.js
```

**Comment by lmiller1990** [comment]
Hi! We noticed something similar using Vite internally at Cypress. For now I'd recommend adding the main libraries to your `viteConfig`, like this:

https://github.com/cypress-io/cypress/blob/develop/packages/app/vite.config.ts#L8-L10

We can look tell Cypress to wait for pre-optimization, but this is likely the best way to get your tests running in a fast, deterministic fashion. Can you give this a try?

**Comment by mjhenkes** [comment]
Right now there doesn't seem to be enough information to reproduce the problem on our end. We'll have to close this issue until we can reproduce it. This does not mean that your issue is not happening - it just means that we do not have a path to move forward.

Please open a new issue with a reproducible example and link to this issue. Here are some tips for providing a [Short, Self Contained, Correct, Example](http://sscce.org/) and our own [Troubleshooting Cypress](https://on.cypress.io/troubleshooting) guide.

**Comment by mcheung610** [comment]
@mjhenkes @lmiller1990 I am running into this bug as well. Was curious what the workaround you provided above?

**Comment by lmiller1990** [comment]
@mcheung610 Vite will do a pre-cache before running. You might not see it often locally, since it caches in `node_modules/.vite`, but on CI, since you likely have a fresh `node_modules` each run, it makes CI seem more unreliable.

What we've done is added all our dependencies to `optimizeDeps.include`. See here: https://github.com/cypress-io/cypress/blob/develop/packages/app/vite.config.mjs#L8-L19

You can clear your vite cache with `rm -rf node_modules/.vite` (or vite may have a command for this).

Vite 3 might have fixed this, though - they address [cold start issues](https://vitejs.dev/blog/announcing-vite3.html#cold-start-improvements) in their release post. I'd recommending updating to Vite 3, then, if you still have issues, try adding the deps required during testing to your config, like we have.

**Comment by espipj** [comment]
Hello @lmiller1990 
I'm coming across this issue...
is it possible that this is a problem Cypress side. I believe the test runner should be able to identify the reloads coming from these vite optimisations and not fail/skip the tests? I don't think Vite 3 fixes this.

**Comment by nagash77** [comment]
Hi @espipj , looks like we are still waiting on a reproducible example of this problem. Can yo provide one using [Cypress Test Tiny](https://github.com/cypress-io/cypress-test-tiny)?

**Comment by Flamenco** [comment]
This is still an issue with the current `Vite` and `Cypress` builds.  It happens when using the Vuetify library for me as well, while doing e2e tests.  I think it is due to using lazy loaded Vue components.

What works for me is adding any items that show up in the `new dependencies optimized` Vite message to the Vite `optimizeDeps` config (as mentioned in previous posts).  In addition, I was not able to use a wildcard `vuetify/lib/components/*`, as that opened up another can of worms with Vuetify.

```
optimizeDeps: {
    include: [
      'vuetify/lib/components/VDatePicker/index.mjs',
      'vuetify/lib/components/VDataTable/index.mjs',
      'vuetify/lib/components/VFooter/index.mjs',
      // and so on.
    ]
  },
```

**Comment by bvandrc** [comment]
This 100% solved it for me: In your Cypress config's Vite dev server config, add this:

```ts
viteConfig: {
...
optimizeDeps: {
   entries: [
      'cypress/support/component.ts',
      'cypress/tests/component/**/*', // or wherever your component spec files are. You could also just move the value you're passing to specPattern outside the config and pass it here as well
   ],
}
```
